### PR TITLE
[Windows] Fix inaccurate 10-bit SDR / HDR color using Pixel Shaders

### LINKED ISSUE #24712: Inaccurate HDR color due to Pixel Shaders being selected as Render Method
### Describe the bug
Here is a clear and concise description of what the problem is:

When Pixel Shaders are selected as Render Method in Kodi Omega Beta 2 running on Windows 11 Pro, HDR colors are visibly and measurably inaccurate. Kodi is selecting Pixel Shaders automatically on my mini PC with AMD Ryzen 5 5560U with Radeon Graphics but uses DXVA on my non-AMD systems. Moreover, it cannot be forced to use DXVA on my AMD system by changing the Render Method setting.

I observed visibly inaccurate HDR gray scale patterns when evaluating my new Beelink SER5 5560U mini PC with integrated AMD Radeon graphics. The patterns exhibit a subtle green tint compared to a Beelink N100 mini PC (Intel UHD graphics), my main PC with Nvidia RTX 3060 Ti, my Chromecast with Google TV (Android 12), and my Sony A80J (Android 10), all of which appear the same in Kodi and measure the same with Calman Home for Sony and my i1D3 meter. SDR is fine on the AMD and all the other devices; it's just HDR that’s the problem, and just on the AMD. My TV is a Sony A80J OLED, and I’m running Windows 11 Pro on all the Windows machines, which are fully updated and using current drivers. For the Windows machines, I have Kodi configured for Limited Range, the video cards for Full Range, and the TV for Limited Range. Black/white clipping patterns are visually fine on all devices, and I calibrated the TV for SDR per standard Sony procedure, yielding dE2000 values < 1. The same calibration works for all my devices.

I always leave Kodi’s Render Method at its default Auto Detect, but comparing debug logs, I found that for the AMD 5560U, Kodi chose Pixel Shaders for my HDR file instead of DXVA like it does for the other Windows machines. This led me to test the normally unaffected Intel N100 by manually selecting Pixel Shaders and DXVA render methods for both HDR and SDR grayscale (white) patterns, and the N100 exhibited the same problem as the 5560U for HDR and Pixel Shaders. I measured the output with Calman and my i1D3, and the dE2000 was 8.1 for that combination, consistent with the AMD measurement, which is quite bad. OTOH, the dE for HDR and DXVA on the N100 was an excellent 0.85, and the dE for SDR was even better, around 0.3 for both DXVA and Pixel Shaders.

In summary, the only bad combination out of HDR/SDR and DXVA/Pixel Shaders is HDR and Pixel Shaders, and it affects both the AMD 5560U and Intel N100 machines. (I didn’t test HDR/Pixel Shaders on my Nvidia card but would assume it’s the same.) The difference is that Kodi selects Pixel Shaders for HDR on the AMD machine when Render Method is set to the default Auto Detect, and trying to override this by choosing DXVA doesn’t help. 

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

On the AMD 5560U machine, Kodi should have switched to HDR and preserved color fidelity like it does on Intel, Nvidia, and Android.

## Actual Behavior

On the AMD 5560U machine, HDR video has a slight green tint.

## Possible Fix

Examining the AMD logs, Kodi is unable to use DXVA and switches to Pixel Shaders for Render Method for HDR. On Intel and Nvidia, it uses DXVA for Render Method. I replicated the problem by configuring Intel Kodi to use Pixel Shaders. So, there is definitely a problem when using Pixel Shaders for HDR, and there may be a problem with Render Method selection on the AMD 5560U, i.e. “Why wasn’t DXVA automatically selected?”

One thing that is not a potential workaround is setting Render Method to DXVA on the AMD system. Kodi ignores the setting and uses Pixel Shaders anyway.

### To Reproduce

1. I’ve attached three video files to this report, and I got them from Mehanik’s HDR10 test file set and the AVS HD 709 set I found here:
        [HDR10 test patterns set | AVS Forum](https://www.avsforum.com/threads/hdr10-test-patterns-set.2943380/)
        [AVS HD 709 - Blu-ray & MP4 Calibration | AVS Forum](https://www.avsforum.com/threads/avs-hd-709-blu-ray-mp4-calibration.948496/)

2. The first file is for visual comparison, and it is fairly easy to see that something is wrong when Pixel Shaders are in use, “Mehanik - 02. C300-C535.mp4", which appears in the linked pattern set under "Mehanik HDR10 test patterns\03. Grayscale\03. Steps\01. Full steps\ ". If your setup is affected by the problem, and you’re familiar with calibrated screens, you might notice a green tint to these grays, as I did. If you have an Intel or Nvidia card, a Chromecast with Google TV or other Android device, try comparing the AMD playback to them:

https://github.com/xbmc/xbmc/assets/101072550/d411ca4a-40ee-4c26-9e88-e9331f6b4426

3. If you have a meter and software, try measuring patterns with DXVA and Pixel Shaders manually selected. I used this pattern for the HDR measurements in my screenshots, "Mehanik - 11. 50%_C504.mp4", which is found in the linked pattern set at "Mehanik HDR10 test patterns\03. Grayscale\01. 21 Step Grayscale Calibration\01. mehanik\01. 10% window 5% steps\01. Calman and HCFR\01. MaxCLL 1000nit\":

https://github.com/xbmc/xbmc/assets/101072550/f63a5d73-c11a-4d3a-8d05-df36ec3fe77f

4. I used this pattern for SDR, "AVS - 01-100% White.mp4", which is found in the linked pattern set at, "AVS HD 709\CalMAN Windows\11 Step Grayscale\":

https://github.com/xbmc/xbmc/assets/101072550/a89e531c-859e-40ad-89aa-b58b86ac4ab6

5. I’ve noted that Kodi on my AMD 5560U machine selects Pixel Shaders as Render Method for HDR when the default Auto Detect is in effect, while Intel and Nvidia choose DXVA. It would be interesting to see if Kodi does the same for other AMD chips, or if it’s peculiar to ones like mine, which I believe is nominally a laptop CPU with integrated Radeon graphics despite it being used in my mini PC.

### Debuglog

The debuglog can be found here:

I’m providing two debug logs for the HDR playback, the first for AMD 5560U, the second for Intel N100. Both have Render Method set to the default Auto Detect. In the AMD log, Kodi switches to Pixel Shaders for the renderer, while it uses DXVA in the Intel log.

1. [AMD 5560U, HDR, Render Method Auto Detect (selects Pixel Shaders - bad)](https://paste.kodi.tv/igipaqesis.kodi)
2. [Intel N100, HDR, Render Method Auto Detect (selects DXVA - good)](https://paste.kodi.tv/teyehehijo.kodi)

### Screenshots 
Here are some links or screenshots to help explain the problem:

These are screenshots from Calman Home for Sony taken after measuring the above patterns. The first two screenshots are HDR and use the same pattern, a 50% gray from Mehanik’s 1000 nit grayscale set, which appears as a somewhat dim white on my TV. The last screenshot is SDR of 100% white window from the AVS HD 709 pattern set. 

AMD 5560U HDR (Render Method Auto Detect with Pixel Shaders selected by Kodi), also Intel N100 HDR (Pixel Shaders manually selected by me) - Bad, dE2000 = 8.132:
image
  
Intel, Nvidia, Android 10, 12 HDR (default Render Method for all, or DXVA manually selected in Windows) - Good, dE2000 = 0.849:
image

SDR for all devices, including AMD, and for the Windows machines, DXVA and Pixel Shaders measure the same - Good, dE2000 = 0.2:
image
 
 

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [x ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 11 Pro
 - Kodi version: Omega Beta 2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by thexai**
## Description
Fix inaccurate HDR color using Pixel Shaders

Fixes https://github.com/xbmc/xbmc/issues/24712

## Motivation and context
In some places we using 8 bit precision for video range full/limited conversion constants and apply on 10 bit / HDR video path.

While in 8 bit limited range low value is 16, high is 235, etc. and scale constants can be defined as:

scale = (235 - 16) / 255;
trans = 16 / 255;

This is not much precise but error is acceptable because is <= that quantization due use 255 steps and no produces visible change. But these constants can not be re-used for 10-bit video because then the error is bigger that quantization using 1024 steps (2 ^10).

Example:

16.0 / 255 = 0.0627450980
64.0 / 1023 = 0.0625610948

(235 - 16) / 255   = 0.8588235294
(940 - 64) / 1023 = 0.8563049853

Is similar but not equally.... so if is used 0.0627450980 constant in 10-bit video this introduces an error.

Potentially are affected all platforms but Android usually only uses MediaCodec path and in Linux (GL) I don't know if this code path is usually used or even if HDR is well supported/used.

These accumulative small errors in video range conversion and YUV to RGB conversion end up producing a slightly visible imperfection on 50% gray with green tint.

NOTE: changes in `output_d3d.fx` are not strictly necessary and not contribute in green tint. Only changed for consistency and probably only introduces small inaccuracy in black level / high level when using limited range output.

## How has this been tested?
Runtime Windows x64 + Nvidia RTX 4070 
Video chain configured in limited range (limited Kodi - full Nvidia - limited TV)

## What is the effect on users?
Fixes inaccurate colors using Pixel Shaders with HDR video. The picture corrected with the PR is consistent with that obtained when using DXVA with the same setup.

## Screenshots (if appropriate):

**Before**
before

rgb-before

RGB (124, 129, 125)

**After**
after

rgb-after

RGB (127, 128, 127)

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by CrystalP** [comment]
The reasoning is sound, I won't be able to runtime test until the weekend.
Some spec documents contain tables of expected values. I don't remember which one off the top of my head but that would be a way to validate a wide range of values in addition to 50% gray.

**Comment by thexai** [comment]
10-bit SDR is also affected:

**master-DXVA**
master-SDR-DXVA

**master-PS**
master-SDR-PS

**PR-DXVA**
PR-SDR-DXVA

**PR-PS**
PR-SDR-PS


Same green tint and now is more clear black level is wrong in SDR:

**black level Pixel Shaders before**
bl-before

**black level Pixel Shaders PR**
bl-after


Black level in limited range is RGB (16, 16, 16) as range is 16 to 235

**Comment by thexai** [comment]
Rebased and fixed also 12-bit

**Comment by thexai** [comment]
I will merge at Friday if nobody has nothing against. I consider have done enough tests myself to be sure the changes are correct. 

Tested on Nvidia RTX 4070 and Intel NUC8i3BEK (that supports DXVA HDR limited range to compare).

File: system/shaders/output_d3d.fx [hunk 1]
```diff
@@ -194,7 +194,7 @@ float4 OUTPUT_PS(VS_OUTPUT In) : SV_TARGET
 {
   float4 color = g_Texture.Sample(KernelSampler, In.TextureUV);
   [flatten] if (m_params.x)
-    color = saturate(0.0625 + color * 219.0 / 255.0);
+    color = saturate((64.0 / 1023.0) + color * (940.0 - 64.0) / 1023.0);
```
Comment (CrystalP) [inline system/shaders/output_d3d.fx:197]: Is it OK for this one not to take the number of bits of the output into account?

**Pull request merged**